apollo-router/src/query_planner/caching_query_planner.rs

```diff
@@ -1,157 +1,165 @@
-{
-    async fn retrieve(&self, key: QueryKey) -> Result<QueryPlannerContent, CacheResolverError> {
-        self.delegate.get(key).await.map_err(|err| err.into())
-    }
-}
-
-#[async_trait]
-impl<T: QueryPlanner + Clone> QueryPlanner for CachingQueryPlanner<T> {
-    async fn get(&self, key: QueryKey) -> PlanResult {
-        let entry = self.cache.get(&key).await;
-        if entry.is_first() {
-            let res = self.delegate.get(key).await;
-            entry.insert(res.clone()).await;
-            res
-        } else {
-            entry
-                .get()
-                .await
-                .map_err(|_| QueryPlannerError::UnhandledPlannerResult)?
-        }
-    }
-}
-
-impl<T: QueryPlanner + Clone + 'static> tower::Service<QueryPlannerRequest>
-    for CachingQueryPlanner<T>
+impl<T: Clone + Send + 'static> tower::Service<QueryPlannerRequest> for CachingQueryPlanner<T>
 where
-    T: tower::Service<
-        QueryPlannerRequest,
-        Response = QueryPlannerResponse,
-        Error = tower::BoxError,
-    >,
+    T: tower::Service<QueryPlannerRequest, Response = QueryPlannerResponse, Error = BoxError>,
+    <T as tower::Service<QueryPlannerRequest>>::Future: Send,
 {
     type Response = QueryPlannerResponse;
-    // TODO I don't think we can serialize this error back to the router response's payload
-    type Error = tower::BoxError;
+    type Error = BoxError;
     type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
 
     fn poll_ready(&mut self, _cx: &mut task::Context<'_>) -> task::Poll<Result<(), Self::Error>> {
         task::Poll::Ready(Ok(()))
     }
 
     fn call(&mut self, request: QueryPlannerRequest) -> Self::Future {
-        let key = (
-            request.query.clone(),
-            request.operation_name.to_owned(),
-            request.query_plan_options,
-        );
-        let qp = self.clone();
+        let mut qp = self.clone();
         Box::pin(async move {
-            qp.get(key)
-                .await
-                .map(|query_planner_content| {
-                    if let QueryPlannerContent::Plan { plan, .. } = &query_planner_content {
-                        match (&plan.usage_reporting).serialize(Serializer) {
-                            Ok(v) => {
-                                request.context.insert_json_value(USAGE_REPORTING, v);
-                            }
-                            Err(e) => {
-                                tracing::error!(
-                                    "usage reporting was not serializable to context, {}",
-                                    e
-                                );
+            let key = (
+                request.query.clone(),
+                request.operation_name.to_owned(),
+                request.query_plan_options.clone(),
+            );
+            let context = request.context.clone();
+            let entry = qp.cache.get(&key).await;
+            return if entry.is_first() {
```

Comment:
I think this return is useless because it's already a tail expression

Reply:
funny, I would have thought that clippy would catch that

Reply:
fixed in fcf34906

Reply:
Clippy should catch that, it's a bug IMO